File: common/src/main/java/org/tdf/common/util/HexBytes.java
package org.tdf.common.util;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.google.common.primitives.Bytes;
import lombok.NonNull;
import org.apache.commons.codec.binary.Hex;
import org.tdf.common.serialize.HexBytesDecoder;
import org.tdf.common.serialize.HexBytesEncoder;
import org.tdf.rlp.RLPDecoding;
import org.tdf.rlp.RLPEncoding;

import java.io.Serializable;
import java.util.Arrays;

/**
 * hex bytes helper for json marshal/unmarshal
 * non-null immutable wrapper fo byte[] inspired by ByteArrayWrapper
 * non-null
 * <p>
 * HexBytes bytes = mapper.readValue("ffff", HexBytes.class);
 * String json = mapper.writeValueAsString(new HexBytes(new byte[32]));
 */
@JsonDeserialize(using = HexBytesUtils.HexBytesDeserializer.class)
@JsonSerialize(using = HexBytesUtils.HexBytesSerializer.class)
@RLPEncoding(HexBytesEncoder.class)
@RLPDecoding(HexBytesDecoder.class)
public final class HexBytes implements Comparable<HexBytes>, Serializable {

    public static final byte[] EMPTY_BYTES = new byte[0];
    // singleton zero value of HexBytes
    public static final HexBytes EMPTY = new HexBytes(EMPTY_BYTES);
    /**
     * generated by vscode
     */
    private static final long serialVersionUID = -1770889227783524732L;
    private final byte[] bytes;
    private final int hashCode;
    private String hexCache;

    private HexBytes(@NonNull byte[] bytes) {
        this.bytes = bytes;
        this.hashCode = Arrays.hashCode(bytes);
    }

    public static String encode(byte[] bytes) {
        return Hex.encodeHexString(bytes);
    }

    public static byte[] decode(@NonNull String hex) throws RuntimeException {
        try {
            return Hex.decodeHex(hex);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public static HexBytes fromHex(@NonNull String hex) throws RuntimeException {
        try {
            return fromBytes(Hex.decodeHex(hex));
        } catch (Exception e) {
            throw new RuntimeException("invalid hex string " + hex);
        }
    }

    public static HexBytes fromBytes(byte[] bytes) {
        if (bytes == null || bytes.length == 0) return EMPTY;
        return new HexBytes(bytes);
    }

    public static HexBytes empty() {
        return EMPTY;
    }

    public byte[] getBytes() {
        return bytes;
    }

    public int size() {
        return bytes.length;
    }

    public boolean isEmpty() {
        return this == EMPTY || bytes.length == 0;
    }

    public String toString() {
        return toHex();
    }

    public String toHex() {
        if (hexCache != null) return hexCache;
        hexCache = Hex.encodeHexString(bytes);
        return hexCache;
    }

    public HexBytes concat(HexBytes another) {
        return new HexBytes(Bytes.concat(bytes, another.bytes));
    }

    public HexBytes slice(int start, int end) {
        if (isEmpty()) return EMPTY;
        while (start < 0) start += bytes.length;
        while (end < 0) end += bytes.length;
        return new HexBytes(Arrays.copyOfRange(bytes, start, end));
    }

    public HexBytes slice(int start) {
        return this.slice(start, bytes.length);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        HexBytes hexBytes = (HexBytes) o;
        return FastByteComparisons.equal(bytes, hexBytes.bytes);
    }

    @Override
    public int compareTo(HexBytes o) {
        return FastByteComparisons.compareTo(
                bytes, 0, bytes.length,
                o.getBytes(), 0, o.getBytes().length);
    }

    @Override
    public int hashCode() {
        return hashCode;
    }
}

File: common/src/main/java/org/tdf/common/util/HexBytes.kt
package org.tdf.common.util

import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import com.github.salpadding.rlpstream.RlpBuffer
import com.github.salpadding.rlpstream.RlpWritable
import com.github.salpadding.rlpstream.StreamId
import com.github.salpadding.rlpstream.annotation.RlpCreator
import org.spongycastle.util.encoders.Hex
import org.tdf.common.util.HexBytesUtil.HexBytesDeserializer
import org.tdf.common.util.HexBytesUtil.HexBytesSerializer

typealias Address = HexBytes
typealias H256 = HexBytes
typealias H2048 = HexBytes

/**
 * hex bytes helper for json marshal/unmarshal
 * non-null immutable wrapper fo byte[] inspired by ByteArrayWrapper
 * non-null
 *
 *
 * HexBytes bytes = mapper.readValue("ffff", HexBytes.class);
 * String json = mapper.writeValueAsString(new HexBytes(new byte[32]));
 */
@JsonDeserialize(using = HexBytesDeserializer::class)
@JsonSerialize(using = HexBytesSerializer::class)
class HexBytes private constructor(val bytes: ByteArray) : Comparable<HexBytes>, RlpWritable {
    private val hashCode: Int = bytes.contentHashCode()

    operator fun get(index: Int): Int {
        return java.lang.Byte.toUnsignedInt(bytes[index])
    }

    val size: Int = bytes.size

    fun isEmpty(): Boolean = bytes.isEmpty()

    override fun toString(): String {
        return hex
    }

    val hex: String by lazy {
        Hex.toHexString(bytes)
    }


    operator fun plus(other: HexBytes): HexBytes {
        return HexBytes(
            ByteUtil.merge(bytes, other.bytes)
        )
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val hexBytes = other as HexBytes
        return FastByteComparisons.equal(bytes, hexBytes.bytes)
    }

    override fun compareTo(other: HexBytes): Int {
        return FastByteComparisons.compareTo(
            bytes, 0, bytes.size,
            other.bytes, 0, other.bytes.size
        )
    }

    override fun hashCode(): Int {
        return hashCode
    }

    override fun writeToBuf(rlpBuffer: RlpBuffer): Int {
        return rlpBuffer.writeBytes(bytes)
    }

    companion object {
        val EMPTY_BYTES = ByteArray(0)

        // singleton zero value of HexBytes
        @JvmField
        val EMPTY = HexBytes(EMPTY_BYTES)


        @RlpCreator
        @JvmStatic
        fun create(bin: ByteArray, streamId: Long): HexBytes {
            return fromBytes(StreamId.asBytes(bin, streamId))
        }

        @JvmStatic
        fun encode(bytes: ByteArray): String {
            return Hex.toHexString(bytes)
        }

        @JvmStatic
        fun decode(hex: String): ByteArray {
            return Hex.decode(if (hex.startsWith("0x")) hex.substring(2) else hex)
        }

        @JvmStatic
        fun fromHex(hex: String): HexBytes {
            return fromBytes(decode(hex))
        }

        @JvmStatic
        fun fromBytes(bytes: ByteArray?): HexBytes {
            return if (bytes == null || bytes.isEmpty()) EMPTY else HexBytes(bytes)
        }

        @JvmStatic
        fun empty(): HexBytes {
            return EMPTY
        }
    }

}